### ISSUE TITLE
INSERT...VALUES...RETURNING simply cannot be trusted long term for multiple rows; implement sentinel columns

### ISSUE THREAD

**Issue by zzzeek** [sqlalchemy/sqlalchemy#9618]
i've been researching this for two days and the answer is the same: yes, the rows seem to be ordered, but no, nobody can guarantee it.   SQL Server is just the one actually breaking the rule really badly right now.    Over on pep-249, we are probably going to even have cursor.executemany() [no longer guarantee the order of records given](https://www.mail-archive.com/db-sig@python.org/msg02071.html).

The form of INSERT (a, b, c) VALUES...RETURNING is certainly one we can embed a hardcoded, numerically incrementing **sentinel** value within.  However syntactically, there is **no way to get it back unless a column is added to the table**.

So I am proposing a new column argument called ``insert_sentinel=True``:

```py
class A(Base):
    __tablename__ = "a"

    id: Mapped[int] = mapped_column(primary_key=True)
    data: Mapped[str]

    _sentinel = mapped_column(insert_sentinel=True)
```

We then would do our INSERTs, when this column is present, in this fashion:

    INSERT INTO a (data, _sentinel) VALUES (?, 1), (?, 2), (?, 3), (?, 4), ... RETURNING a.id, a._sentinel

then we sort by _sentinel.  the insertion of the "sentinel" is already done in https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4548.  

Among other things this will be a real solution to the problem faced at #9603 and will also be a hedge against any other database where we can't guarantee INSERT ordering.

for use_insertmanyvalues, it then becomes a three-state flag, like "use_insertmanyvalues='sentinelonly'" vs. "use_insertmanyvalues=True".   We would have to warn for tables that don't have sentinels (either that we can't use insertmanyvalues and things are too slow, or that we are using insertmanyvalues but the table is not "safe").

all the use/ warnings for "sentinel" **only** apply if we are trying to use RETURNING.  if we arent RETURNING, there is **no problem** - that is, we really (we could do this anyway) can turn on insertmanyvalues for SQL Server as long as the statements aren't using RETURNING.

The use of "insertmanyvalues" with RETURNING is mostly an ORM use case.   You can use this directly in 2.0 but it's the ORM that really relies upon it.  in ORM use cases, people are usually using the ORM to manage their tables.

Whether or not we make this "required" or whatever, it seems we at least should have this capability available.    

It's also supported in all databases, even SQLite, to ALTER TABLE now that simply adds a nullable column.  So even for systems that use pre-existing databases, we could have functions that do something like ensure_sentinel(connection, table) that just adds this column if not already present; that is it doesn't even even need to be within the usual migration path.  it's a simple nullable integer (or even smallint) column, only needs to store a number as big as the actual batch size, does not need indexes, does not need to be SELECTable, nothing.  it is only used in INSERT as a counter.

**Comment by zzzeek** [comment]
see https://twitter.com/mitsuhiko/status/1644400386977955840

cc @mitsuhiko

**Comment by zzzeek** [comment]
I think we should have a very quick device to indicate this for tables, etc. and it would just have to be all over the docs:


```py
Table(
    't', metadata,
   Column('id', Integer, primary_key=True),
   Column("data", String(50)),
   insert_sentinel()
)
```

Declarative:


```py

class Base(DeclarativeBase):
    _sentinel = insert_sentinel()

```

**Comment by CaselIT** [comment]
I don't really like this, but I can't find a better solution that does not involve server side programming. 

edit: ok on pg I found a way but is idiotic and it probably no better than an execute many:
```sql
with 
c0(id) as (insert into t(val) values (42) returning t.id),
c1(id) as (insert into t(val) values (43) returning t.id),
c2(id) as (insert into t(val) values (44) returning t.id),
c3(id) as (insert into t(val) values (45) returning t.id),
c4(id) as (insert into t(val) values (46) returning t.id)
select c0.id from c0
union all select c1.id from c1
union all select c2.id from c2
union all select c3.id from c3
union all select c4.id from c4
```

**Comment by zzzeek** [comment]
only if these databases started helping us, like adding a magic returning value "rowid" or something.  but we'd have to appeal to every DB vendor and it would take years

**Comment by CaselIT** [comment]
btw pg does not seem to have a low limit on ctes, so the idiotic solution "works" but it takes a lot of time, probably parsing/planning: for 2000 entries: `Planning Time: 869.175 ms Execution Time: 83.891 ms`

**Comment by zzzeek** [comment]
the only question is how do we roll this out:


1. turn off insertmanyvalues across the board for all dialects that dont have "sentinel"?  
2. if number 1, do we 
     a. warn when an INSERT RETURNING happens that performance is degraded or 
     b. do nothing, and everyone gets degraded performance out of the gate for 2.0.10 with no warning?
     if b. do we also c. do we have to reinstate the FAQ entry at https://docs.sqlalchemy.org/en/14/faq/performance.html#i-m-inserting-400-000-rows-with-the-orm-and-it-s-really-slow , we were done with "the ORM is slow!" but now we are not anymore
3. if not number 1, do we :
    a. warn when an INSERT RETURNING happens that we can't 1000% guarantee future versions of PG / SQLite / MariaDB will work , please use "sentinel" or
    b. do nothing?

**Comment by zzzeek** [comment]
here's Ruby on Rails doing it for Postgresql:

https://github.com/rails/rails/blob/ae941226c41a72dacc1eb536a3fbf937196fa33e/activerecord/lib/active_record/persistence.rb#L71

**Comment by zzzeek** [comment]
OK here's another thought.

we do actually have the same feature as what rails has above, in our [bulk insert](https://docs.sqlalchemy.org/en/20/orm/queryguide/dml.html#getting-new-objects-with-returning) feature.

there is actually *no problem* with our current insertmanyvalues implementation for the above feature.  because nothing about what we have there is assuming anything about the order of RETURNING, there is not actually any existing object that's being matched up.

Which means we really only have to scale back insertmanyvalues, *in the unit of work*.    Also for SQL Server, I pulled out "insertmanyvalues" entirely even for non-returning cases which is overkill.

since it seems like all the DBs we're going to have to worry about this, we can:

1. keep insertmanyvalues the same, but also add support for "sentinel" as above
2. restore it for SQL Server, since it "works" just fine if you dont care about RETURNED order
3. add *another* dialect function used **only by the unit of work** which is something like "has_sentinel" or "insertmanyvalues_deterministic_returning(table)", something like that
4. limit the use of insertmanyvalues **just in the unit of work**.   after all that was the issue we had in #9603

**Comment by sqla-tester** [comment]
**Mike Bayer** has proposed a fix for this issue in the **main** branch:

**implement sentinel column** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4548

**Comment by CaselIT** [comment]
> the only question is how do we roll this out:
> 
>     1. turn off insertmanyvalues across the board for all dialects that dont have "sentinel"?
> 
>     2. if number 1, do we
>        a. warn when an INSERT RETURNING happens that performance is degraded or
>        b. do nothing, and everyone gets degraded performance out of the gate for 2.0.10 with no warning?
>        if b. do we also c. do we have to reinstate the FAQ entry at https://docs.sqlalchemy.org/en/14/faq/performance.html#i-m-inserting-400-000-rows-with-the-orm-and-it-s-really-slow , we were done with "the ORM is slow!" but now we are not anymore
> 
>     3. if not number 1, do we :
>        a. warn when an INSERT RETURNING happens that we can't 1000% guarantee future versions of PG / SQLite / MariaDB will work , please use "sentinel" or
>        b. do nothing?

There is no easy answer here. My current thinking is
1 only for unit of work, other uses are ok, but we should what on the docs the the order of insert returning is "lol whatever"
2 b and c. I guess we have to dust off the bulck insert things 
3 b. As mentioned above we may warn in the docs, but I don't think a warn in code is not needed.


I thought of another possible alternative that's relaxing some of the guardees given by sqlalchemy:
Assumption: in most cases then this feature actually makes a performance difference the user code does not care about the object instances once they are added to the db.
Idea: add an method called something like add_unordered that basically does the same as bulk insert and then explires all values added. At that point we can assign whatever pk to the list of objects since they are all expired and we are saying that we are not respecting order. Or even instead of expire we could reload the objects using `returning *`
Drawback: complicates the session and unit of work quite a bit, there is already bulk insert that basically does this

**Comment by CaselIT** [comment]
A further option that it can be implemented is using unique contraints that are non nullable and integer or uuid or text type (if we assume that case collation ect are comparable with python). 

Then we could use a dict to match the returning rows with the values from the rows.

Not sure it's worth it though

**Comment by zzzeek** [comment]
this is a good point that we might want to consider marking an existing column as a "sentinel" but it gets difficult to ensure that the col always has values for INSERT which we can locate in the params .

as far as add_unordered() that would not be worth it and also people might have other non-ORM state on their objects as well as existing associations, I dont think it would be workable.     as we see in the example we started with,it had to do with related objects being coordinated correctly.

**Comment by CaselIT** [comment]
> gets difficult to ensure that the col always has values for INSERT which we can locate in the params .

if we limit ourselves to unique non null columns you kind of need to have a value there, or the db will complain

> as far as add_unordered()

I agree, it seems a more complex way of implementing the existing bulk inserts


In any case I think it would be still useful to ask the various db communities about this, maybe some implementation actually do guarantee the order of things. At the very least they could suggest things that we have not though about.
I'll try posting some questions and report back

Returning to the topic, maybe the default name should be something like `_sa_sentinel` or `_sa_insert_counter` on the db side. Also I think it would be better to use a different constructor, not `mapped_column(insert_sentinel=True)`, since we would like in general to ignore this value in inits (and maybe error if set by the user) and using `mapped_column(...)` requires setting `init=False` when using dataclasses, while a different construct may be automatically ignored by dataclasses  and type checkers

**Comment by zzzeek** [comment]
yeah i thought we'd have a specific construct insert_sentinel()

the underscore name is a little torublesome because some dbs consider that to be a quoted name (Oracle...which...doesnt need a sentinel column)... still works anyway

**Comment by CaselIT** [comment]
For oracle we could even make that an hidden column, so it doesn't clutter `select *`

**Comment by CaselIT** [comment]
For sqlite there is no need to ask. The documentation is fortunately pretty explicit:

> While SQLite does guarantee that all database changes will occur before any RETURNING output is emitted, it does not guarantee that the order of individual RETURNING rows will match the order in which those rows were changed in the database. The output order for the RETURNING rows is arbitrary and is not necessarily related to the order in which the rows were processed internally. 

and

> The rows emitted by the RETURNING clause appear in an arbitrary order. That order might change depending on the database schema, upon the specific release of SQLite used, or even from one execution of the same statement to the next. There is no way to cause the output rows to appear in a particular order. Even if SQLite is compiled with the [SQLITE_ENABLE_UPDATE_DELETE_LIMIT](https://www.sqlite.org/compile.html#enable_update_delete_limit) option such that ORDER BY clauses are allowed on DELETE and UPDATE statements, those ORDER BY clauses do not constrain the output order of RETURNING.

from https://www.sqlite.org/lang_returning.html

**Comment by zzzeek** [comment]
awesome

**Comment by CaselIT** [comment]
I plan to search for recent reference and ask on pg if I find nothing.

At the very least they could add that to the docs, since it seems like something it's missing.

Same for the mariadb docs, where they don't even mention order once

**Comment by mitsuhiko** [comment]
For what it's worth I would never have assumed that this would come back in any specific order by *any* SQL implementation. Reason being that unless you use `ORDER BY` (which you can't do here, unless you use CTEs) the result order is undefined. I also cannot find anything in postgres in particular that would (even accidentally) return the results in insertion order, so I'm fairly confident that if vaccums are happening or concurrent deletes on large inserts, you would get the tuples back in arbitrary order today already.

**Comment by CaselIT** [comment]
I think we assumed this since it's what make the most sense, and it's the observed behaviour on postgresql. I agree that we should haven't assumed it.

The very sad thing is that there isn't any way (that we found) of getting back the position of a row in the insert values, without using a sentinel column in the table.
That's the main thing that's missing and I guess part of the reason for assuming that the order were kept

**Comment by zzzeek** [comment]
> For what it's worth I would never have assumed that this would come back in any specific order by _any_ SQL implementation. Reason being that unless you use `ORDER BY` (which you can't do here, unless you use CTEs) the result order is undefined. 

using CTE with ORDER BY brings up the particular thorny thing about this problem which is that there are **two** orderings that are undefined:

```py
WITH foo AS (
 INSERT INTO table (a, b, c) VALUES (%s, %s, %s), (%s, %s, %s), (%s, %s, %s) RETURNING table.id
SELECT id FROM foo ORDER BY id
```

The ORDER BY above guarantees we receive the rows *in primary key order*, which if using SERIAL / IDENTITY would be **insertion order**.  However it does not guarantee that **the insertion order matched the order of the VALUES tuples**.

With MS SQL Server I think I've seen it behaving inconsistently on **both** sides of this problem though it's very hard to figure out as it seems to have a fair degree of heisenbehavior going on.


> I also cannot find anything in postgres in particular that would (even accidentally) return the results in insertion order, so I'm fairly confident that if vaccums are happening or concurrent deletes on large inserts, you would get the tuples back in arbitrary order today already.

I would assume you've spent time with the PG source code to make this claim, as someone would have to be really deeply familiar with the internals to come up with this - but then adding "even accidentally", the evidence is not borne out by this, it clearly maintains the ordering in the "default accidental" case.  We've found many posts of people observing this behavior and asking if it's guaranteed, with only one actual authoritiative [post](https://www.mail-archive.com/pgsql-hackers@postgresql.org/msg204161.html) saying "no, it's not guaranteed" (the same post also confirms that it is most definitely the observed behavior though).     Please note, we are **not** talking about ON CONFLICT, or "INSERT FROM SELECT", or anything like that; strictly INSERT with literal tuple-valued entries directly in the statement and nothing more.

My current take on PostgreSQL is that the behavior is "accidental" and therefore could change at any time but for the time being is the actual behavior.    SQLAlchemy 1.4 was first released in November of 2020 and has been using this approach with PostgreSQL / psycopg2 the whole time, where the ORM unit of work does INSERT..VALUES (), (), (), RETURNING and the UOW then matches those returned rows to the objects by primary key, meaning if that order is off, any situation where those objects have further collections on them we'd be seeing the issue like we see in #9603.    That's tens of millions of downloads per month where we would assume applications have tolerated concurrent VACUUMs and things like that.

If you can identify a case in current PG releases where such an INSERT doesn't succeed that would be helpful because that will determine what we do for the 1.4 series.

**Comment by zzzeek** [comment]
Tom Lane confirms in that thread (in 2012) that the current implementation **does** behave that way:

> I don't believe this is a good idea in the slightest.  Yeah, the current
implementation happens to act like that, but there is no reason that we
should make it guaranteed behavior.  Nor is a regression test case going
to stop someone from changing it, anyway.

now, he is saying, "oh this can change", sure. 

I **guarantee you** they would not be able to change this without a giant storm of complaints because people have been coding to this assumption for a long time.

**Comment by zzzeek** [comment]
that's a truly fascinating thread but through all of it, they forgot to act upon the most important [message](https://www.mail-archive.com/pgsql-hackers@postgresql.org/msg204409.html) in the thread which is, emphasis added:

> I always thought it *was* guaranteed, and I've encountered code written
by other people who were obviously under the same impression: take some
strings (e.g. flag names), use "insert … returning id", map the ids back
to the names, and use the values in further inserts into other tables
("flag_id foreign key references flags").

> I know one could say "returning id, name", but there's certainly code
out there that doesn't do this.

> I personally think the return order should be guaranteed; and if not,
then the documentation **urgently needs some prominent warnings** to tell
people that they should not assume this (for any variant of RETURNING).


I'm not in a hurry to sign up for this list but **WHY** can't they document this warning?

**Comment by zzzeek** [comment]
> Thanks for the update @jdhowroyd
> 
> Using select order by may be an option if a table has an identity column as primary key on mssql. But maybe it could be a follow on optimization? Not sure if you want to try using that @zzzeek in the patch? At least we could ensure that a dialect is able to do something like this?

I spent several days trying every possible SQL Server syntax and im not sure to the degree that the issue was / was not reproducible with each one.      I guess the idea here is yes, the RETURNING is still mis-ordered but at least the statement can be organized so that the INSERTs are in some sortable order which can then be used as the "insert_sentinel" (im likely changing that name).  As mentioned in my other comment, there's two sides to the problem, there's the INSERTs themselves then there's the RETURNING.

It's not clear to me the degree to which we have such syntaxes on other backends, so I would need these syntaxes with documented guarantees of INSERT ordering for SQLite, PG, Mariadb.   Doing things like triggers and stuff are non-starters on this end, if you're going to add a trigger, just add a simple integer column, seems much simpler.

In any case, not all tables have integer primary keys with a sequence so the "hard" approach I'm doing now has to happen first.   A majority of our users are on PG, SQLite, MariaDB will be faced with new warnings for "sentinel columns", which I think I am going to name something like "insert cookies".   There's a lot of work mostly reorganizing as much as possible in the ORM to not need this ordering in the first place.

**Comment by jdhowroyd** [comment]
Thanks!  It seems to be a real can of worms.  My head line was to suggest passing on the decicion to the SQLAlchemy user by providing the mechanisms "to set policy" on a case by case basis (dependant on schema design): but undoubtedly, easier said than done.  Or to put it another another, I seriously doubt you will get cast iron guarentees from the database maintainers / vendors (even for current implementations let alone future developments) that would cover every possibility.

By the way, may I also point out another GOTCHA with mssql.  If there is an error while performing the insert the output clause will return results for the rows inserted until the error occurs, but no actual data will be written to the table (the transaction is rolled back in it's entirety).  So if inserting 50 rows and there is a data conflict on 49th then you can expect 48 returns (from the output clause) and an error message: none of the rows have been written.  I hope I am not pointing out the obvious but thought it worth saying just in case.  No idea what other DBMSs will do here except that if an error is present then the whole transaction should be rolled back (whether or not they output any return values would be architecture specific in my opinion).

**Comment by jdhowroyd** [comment]
@zzzeek:

> I spent several days trying every possible SQL Server syntax and im not sure to the degree that the issue was / was not reproducible with each one. I guess the idea here is yes, the RETURNING is still mis-ordered but at least the statement can be organized so that the INSERTs are in some sortable order which can then be used as the "insert_sentinel" (im likely changing that name). As mentioned in my other comment, there's two sides to the problem, there's the INSERTs themselves then there's the RETURNING.

Yes, absolutely.  Personally, I do like the term sentinel (sounds good to me but you are the coder).

Sorry for any confusion on the sequences: it's just how PostgreSQL does it's identity columns.  I'm not suggesting for one moment that SQLAlchemy needs create those on the fly.  The main point was there may be other ways to do it dependant on schema design.

**Comment by CaselIT** [comment]
In sqlserver the documentation seems to indicate that they do guarantee that if id is an identity the two will align, meaning that the output will be randomly ordered, but if you order by id column you get the same order as the one in the select with that set of values and that order by clause.

I can't find any such text on PG or SQLite, so from them we cannot expect anything here.

>  A majority of our users are on PG, SQLite, MariaDB will be faced with new warnings for "sentinel columns", which I think I am going to name something like "insert cookies".

Personally I don't like insert_cookies, since I fear will cause confusion with the usual meaning of cookies, and to me is unclear what it's use is from the name only. Sentinel to me is clearer here.
Other alternatives I'm come up are "insert progressive" or "insert batch order". These would also make fine column name "_sa_insert_progressive" or "_sa_insert_batch_order".

**Comment by jdhowroyd** [comment]
I was searching on web specifically with regard to PostgreSQL and came across this post on stackoverflow:
https://stackoverflow.com/questions/50809120/postgres-insert-into-with-select-ordering

One of the answers: specifically states "Postgres, explicitly as of 9.6, will logically insert in the order of the returned result set." and points to the code commit:

> The behavior is explicitly codified in this commit: https://github.com/postgres/postgres/commit/9118d03a8cca3d97327c56bf89a72e328e454e63

I must admit documentation else where seems weak.  I also presume "logically inserted" should be treated in the same way as mssql (i.e. generation of ids will be correct, but I don't see anything on what order they will be returned in).

**Comment by zzzeek** [comment]
sqlite we see a similar question, and the usual lecturing from someone who is not the core maintainer

https://sqlite.org/forum/forumpost/ba238d772f

**Comment by CaselIT** [comment]
Since the complete name is not that long we could just call these `insert values sequence number`, with or without "values". (not sure if values is better than batch here). As for the table column something like `_sa_insert_values_sequence_num` should fit in all dbs.

If we fear that sequence may be misinterpreted as having anything to do with a db sequence, we could replace it with order, so `insert values order number` (or just `insert values order` even)

**Comment by jdhowroyd** [comment]
My first reading of that PostgreSQL commit (by no means definitive) is that it does what it says on the tin.  Namely, guarentees that volatile functions are executed after sorts and limit bys; so giving us the correct evaluation of the indentity functions.  I don't see anything about the order of the RETURNING clause there (so a much bigger job to track that down).  Reading between the lines (so hypothesis rather than fact) it seems that PostgreSQL is marshalling the returns so are likely to be in the correct order.  Perhaps this should be a direct question to the maintainers, there is a link to a form for such things at the end of the documentation page on insert:
https://www.postgresql.org/account/comments/new/15/sql-insert.html/
(you may need to create an account with them).  It's probably better coming from you as you'll be closer to the truth.

I also found the following paragraph on the sql select page.

> According to the SQL standard, the expressions in the output list should be computed before applying DISTINCT, ORDER BY, or LIMIT. This is obviously necessary when using DISTINCT, since otherwise it's not clear what values are being made distinct. However, in many cases it is convenient if output expressions are computed after ORDER BY and LIMIT; particularly if the output list contains any volatile or expensive functions. With that behavior, the order of function evaluations is more intuitive and there will not be evaluations corresponding to rows that never appear in the output. PostgreSQL will effectively evaluate output expressions after sorting and limiting, so long as those expressions are not referenced in DISTINCT, ORDER BY or GROUP BY. (As a counterexample, SELECT f(x) FROM tab ORDER BY 1 clearly must evaluate f(x) before sorting.) Output expressions that contain set-returning functions are effectively evaluated after sorting and before limiting, so that LIMIT will act to cut off the output from a set-returning function.

See: https://www.postgresql.org/docs/current/sql-select.html

Followed by a warning that this does not apply for versions of PostgresSQL before 9.6.

Again, I hope that helps.

**Comment by CaselIT** [comment]
Re warning: let's allow the driver to set a threshold value for single orm inserts. defaults to 1000, we can do 1e20 for sqlite to remove it entirely (or just allow negative numbers / None / 0 to disable)

This way it should trigger in a lot less cases and in the warning we could mention also the possibility of doing bulk insert instead of relying on the unit of work

@jdhowroyd I think to get the same behaviour as described in the sql server docs regarding identity you need to run a statement like this:
```sql
INSERT INTO tbl(id, a, b) SELECT nextval(pg_get_serial_sequence('tbl', 'id')), a, b
FROM (values ('a1', 'b1', 1), ('a2', 'b2', 2)) AS vv(a,b, num)
ORDER BY num RETURNING id
```
That is, jut need to compute the identity value in the select and insert it to the db.
If you omit id from the insert list and let the db generate it, you get possibly random order of rows, since they don't guarantee at all what's the order in which the row are added to the db.

The above is not required on sql server according to the docs

**Comment by jdhowroyd** [comment]
> @jdhowroyd I think to get the same behaviour as described in the sql server docs regarding identity you need to run a statement like this:

INSERT INTO tbl(id, a, b) SELECT nextval(pg_get_serial_sequence('tbl', 'id')), a, b
FROM (values ('a1', 'b1', 1), ('a2', 'b2', 2)) AS vv(a,b, num)
ORDER BY num RETURNING id

Thanks for pointing that out.  The test cases in the PostgresSQL commit are of this form (with a specific nextval('sequence') evaluation in the select).  So much more digging if implemented with a trigger...

**Comment by jdhowroyd** [comment]
> @jdhowroyd for SQL server I have this, can you confirm this is the form that's guaranteed to apply IDENTITY to the rows in the order they are placed in the VALUES clause?

INSERT INTO a (data) OUTPUT inserted.id, inserted.id AS id__1 SELECT p0 FROM (VALUES(?, 0), (?, 1), (?, 2), (?, 3), (?, 4), (?, 5), (?, 6), (?, 7), (?, 8), (?, 9), (?, 10), (?, 11), (?, 12), (?, 13), (?, 14), (?, 15), (?, 16), (?, 17), (?, 18), (?, 1 ... 9672 characters truncated ... , (?, 995), (?, 996), (?, 997), (?, 998), (?, 999)) AS imp_sen(p0, sen_counter) ORDER BY sen_counter

Not sure I quite understand, there are two inserted.id listed in the OUTPUT.  So

INSERT INTO a (data) OUTPUT inserted.id AS id__1 SELECT p0 FROM (VALUES(?, 0), (?, 1), (?, 2), (?, 3), (?, 4), (?, 5), (?, 6), (?, 7), (?, 8), (?, 9), (?, 10), (?, 11), (?, 12), (?, 13), (?, 14), (?, 15), (?, 16), (?, 17), (?, 18), (?, 1 ... 9672 characters truncated ... , (?, 995), (?, 996), (?, 997), (?, 998), (?, 999)) AS imp_sen(p0, sen_counter) ORDER BY sen_counter

would make sense to me given a table:

create table a (
 id int identity,
 data whatever
)

and parameters passed to the insert to fill in data.

SQL Server will OUTPUT the id values created (in some arbitrary order), but once sorted could be written back ORM side.

**Comment by zzzeek** [comment]
the inserted.id is there twice yes but that's an implementation artifact, shouldn't actually matter

**Comment by jdhowroyd** [comment]
@zzzeek Should be fine.

**Comment by CaselIT** [comment]
Question on PG mailing list; https://www.postgresql.org/message-id/CAN19dycS0oUorLOGgjxMX8NZq-hOA1RJrzwC4L3FxpMf4BQbBQ%40mail.gmail.com

**Comment by mijoharas** [comment]
Hello, sorry for posting on an old closed issue.
I read the thread [mentioned above](https://www.postgresql.org/message-id/flat/6644419d-50e1-97c9-5507-b6915bec96fa%40gmail.com#9d6427c1ef7a3246a86f4acaf8281535) on the postgres mailing list.
Reading the last few messages it sounded like a reasonable idea was arrived at and a proposal was being formed (`VALUES (..) WITH ORDINALITY` or similar).
I couldn't see any follow up there, could I ask if there were any further mailing list threads on the feature, and whether it was formally proposed in any way? Is there anything that can be done to help that process? I'd also be very keen to see this feature implemented (for the same reason, to link up values inserted with in memory objects.)
Sorry, if this forum isn't the right one to ask in as it's specifically a postgres feature, but given it linked to this issue, and this is what spawned the discussion I thought it was an appropriate place to ask.

**Comment by CaselIT** [comment]
Hi @mijoharas 

No problem regarding asking it here!

> I couldn't see any follow up there, could I ask if there were any further mailing list threads on the feature, and whether it was formally proposed in any way? 

Not that I'm aware, the general interest from some of the developers seemed to be from low to negative, so nothing more come of it at least from my part.

> Is there anything that can be done to help that process? I'd also be very keen to see this feature implemented (for the same reason, to link up values inserted with in memory objects.)

I don't know, sorry. As mentioned it seemed that either I wasn't able to explain why this feature would be useful and why what's currently available is lacking or that use case was not deemed of particular interest / importance. Or at least that was my impression

By all mean feel free to follow up on it. Feel free to link follow ups and or new threads here too

**Comment by jdhowroyd** [comment]
Hi Michael et al.,

Apologies for the tardy reply, I have been really busy with a new job.
Also, for the record, I'm just a developer and nothing to do with
SqlAlchemy except very much a supporter of their excellent work.

I had intended to start on a PostgreSQL patch for this before now, but I
have not had the time or even made a single key stroke to that end (did I
mention the new job).  Anyway, I would be very keen to contribute whatever
I can (in support or collaboration) and if nothing happens I will pick this
up again (in the new year might be a possibility).

Personally, I think this is a really important project and to some extent
the database community has lost its way a little (echoing Mike Bayer's
comments).  A DBMS needs to act as an interface between the programmatic
environment and the actual persistent datastore (original intents).  SQL
has moved on with multirow inserts (update and deletes) and maintaining
state between data programming side and database side has not been
maintained.  This is a classic case of Object Relational Impedance
mismatching (and a little troubling that the folk at PostgreSQL haven't
acknowledged the problem), and actually nothing to do with OOP: however one
chooses to code this is still a problem.  I have a hunch that when
PostgreSQL folk do get around to implementing parallelised inserts, a lot
of people's code will break.

Personally, I am not 100% convinced with the "WITH ORDINALITY" syntax and
suspect a little conflation between concerns.  There was a suggestion of
bringing in a new SQL element "PASSING" (as will be named here).  This
seems to be a more robust way to do things and more conceptually
orthogonal.  My suggested syntax would be as follows.

INSERT [INTO} <table> (<col_1>, ..., <col_m>)
SELECT <val_1>, ..., <val_m>
FROM <query>
[ORDER BY <ord_1>, ..., <ord_k>] -- for SELECT
PASSING (<val_m+1>, ..., <val_m+j>) [[AS] <name>]
RETURNING <ret_1> , ..., <ret_n>
[ORDER BY <out_1>, ..., <out_i>] -- for RETURNING

Here, <table> has (<col_1>, ..., <col_m>) as labelled columns but may have
more with default values or functions (for computing the missing values not
specified, using NULL as the ultimate fallback).  All of these values (inc.
those defaulted or computed) may be accessed in the "RETURNING" clause
using established naming conventions.  There is the natural one to one
mapping of <val_1>, ..., <val_m> to <col_1>, ..., <val_m>.  Any defaulted
or computed columns of the insert are processed in the order of the select
statement (thus giving control of how the assignment of these is performed
(e.g. for sequences)).  The "PASSING" clause may access any columns of
<query> that would be (directly) accessible at the level of the preceding
SELECT <val_1>, ..., <val_m> (if it can be selected then it can be
passed).  These values do not directly affect the data inserted by are
linked to each row inserted as auxiliary values which may then be
returned.  The "RETURNING" clause may access any fields inserted into
<table> (including defaults and computed values) together with the passed
values from the "PASSING" clause, with a guarantee that passed values
correspond to the "outputted" insert values.  While it seems unnecessary to
repeat columns from "PASSING" in the "RETURNING", I would favour this as it
allows specifying the order and column names of those values in the
"RETURNING" clause (however, completely pointless to specify a field as
PASSING and then not accessed in RETURNING, so maybe something to see what
the consensus is).  Finally, the "RETURNING" fields should be marshalled
and sorted should the returning "ORDER BY" clause be given (if not, then
arbitrary (aka most efficient) return order can be used).  I hope the
intent is clear but happy to discuss further.

There are other syntactical elements often used with general insert
statements but I think they all fit in naturally.  This (I hope) purely
expresses the requirements of a "PASSING" clause as an additional element
of SQL syntax to meet the requirements of multirow INSERTS. I would imagine
that a "PASSING" clause also has use with UPDATES, DELETE and MERGE, but I
have not fully sketched these out (the idea is the same for returning
fields from initial queries through to an output in a RETURNING statement
"leapfrogging" the internal table writes.  This makes the effort on the
part of the DBMS explicit in the query (one of the definite requirements
from the PostgreSQL folk).

This is quite involved and touches various aspects of the DBMS including
the language passers, query planning, query optimisation, low level details
of the table writes and the mechanism for linking initial query to final
output.  One of the problems is (at least when I looked a few months ago)
there is no real need to do all this as RETURNING preserves order of the
initial query.  The PostgreSQL folk stated that they are not willing to
commit to this as they are intending to implement parallelisation for
INSERT, UPDATE, DELETE (and presumably MERGE).  I am unclear what
optimisation strategies they currently employ.  For example, MS SQL Server
varies the order of INSERT statements to reduce overhead on constraint
checking (for example, FOREIGN KEYS): so by grouping rows together
constraints can be evaluated (and then forgotten about) reducing the
overall load of the query.  One can, of course, cache constraint checks but
then this needs to be written to memory pages and potentially out to disk
for very large queries.  At the very least, some form of query optimisation
(if only mock examples) need to be considered to implement the proposed
"PASSING" clause and (at least in sketch form) how it would be implemented
with parallelisation.

So by no means a small or trivial project, and one would learn (at least I
would) loads about the internals of PostgreSQL (and DBMSs in general).  As
already stated, I am very keen to contribute in whatever way would make
sense.  My biggest problem is how much time I can commit (at least in the
near future) but I am hoping this should ease in the new year.  Just let me
know, and I'll see what I can do.

I too am not aware of other threads or anybody else taking this forward.

Happy hunting and kindest regards,
John.
